man/showres.1
.TH "showres" 1
.SH "NAME"
showres \- show all reservations
.SH "SYNOPSIS"
.B showres [options]
.SH "DESCRIPTION"
.TP
This program lists all reservations.
.SH OPTIONS
.TP
.B \-d \--debug
Turn on communication debugging
.TP
.B \-h \-\-help
Displays the usage and a brief descriptions of the options
.TP
.B \-\-version
Displays client revision and Cobalt version
.TP
.B \-l
Display shortened reservation listing
.TP
.B \-\-oldts
Use old timestamp
.TP
.B \-x
Print reservations really verbose
.SH "SEE ALSO"
.BR setres (8)
.BR releaseres (8)
.SH "BUGS"


CHANGES
= Changes from previous Cobalt Versions =

== Changes to 0.99.29 ==
 * Interactive jobs are experimentally available for cluster systems.

== Changes to 0.99.28 ==
 * Cobalt now supports qsub options in script mode jobs similarly to other
   job scheduler scripts.  The form is #Cobalt <qsub option> <option argument>.
   Most qsub options are supported, except for mode.  Jobs using this feature must
   be script mode jobs.  This is similar in style to PBS-directives. (Cobalt Ticket 354)
 * Fixed a bug in cluster systems where a failure duing job startup could result
   in all nodes for a job being stuck in the allocated state (Cobalt Ticket 662)
 * $COBALT_STARTTIME and $COBALT_ENDTIME added to job's environments, these are
   the Unix timestamp for when the job started, and when Cobalt will terminate
   the job due to exceeding walltime. (Cobat Ticket 593)
 * A requested walltime of 0 will set the job to run for the time remaining in
   a reservation. (Cobalt Ticket 474)
 * The --env flag to qsub now aggregates multiple instances of the flag rather
   than overriding the prior instance (Cobalt Ticket 657)

== Changes to 0.99.27 ==
 * Alternate kernel support for BG/Q compute nodes and IO nodes is available.
  Numerous configuration settings have been added to support this.
  (Cobalt ticket 621)
 * If $jobid is included in a job name specification or an environment variable,
  at submission time, Cobalt will substitute the generated jobid for that job.
  See man qsub for more details. (Cobalt ticket 617)
 * The Cobalt jobid is now explicitly added in the cobaltlog file. (Cobalt ticket 390)
 * The name of a job can now be specified seperately from naming the output
  files. (Cobalt ticket 625)
 * Multiple locations can be passed to reservations through the -p option as a
  ':'-delimited list. (Cobalt ticket 405)
 * cqadm can now be used to release user hold, as well as put jobs into a user hold
  as well as an admin hold (Cobalt ticket 382)
 * The users field can be removed from a reservation using setres -m -u '*'. (Cobalt
   ticket 381)
 * Releaseres should now report the correct partiton count on reservation release
  (Cobalt ticket 470)
 * Reservations now show time remaining in showres once the reservation is active.
  (Cobalt ticket 585)
 * Schedctl should handle flipping score and jobid arguments more gracefully and 
  no longer silently discards invalid options. (Cobalt ticket 502)
 * ':' and '=' may now be escaped as '\:' and '\=' in the --envs argument (Cobalt
  ticket 589)
 * Reservations may now take 'now' for the start time of a reservation (Cobalt
  ticket 546)
 * "GEOMETRY", "ION_KERNEL" and "ION_KERNELOPTIONS have been added to the JOB_DATA
  table of the Cobalt database 


== Changes to 0.99.26 ==
 * Corrected a bug that was causing messages in the clients that used to go to
  stdout to go to stderr instead

== Changes to 0.99.25 ==
 * Fixes for clients that were causing erronious default values for proccount
  have been fixed.
 * The setgid wrapper now calls a site-selectable python interperter with the
  '-E' flag by default.
 * A bug that could cause TimeRemaining to show negative values has been fixed
 * Queues may be added or removed from blocks without overwriting the entire
  list with new options in partadm using --rmq or --appq with the --queue option.


== Changes to 0.99.24 ==
 * Reverting client changes.  Found some issues when deploying to test systems
 * --defer flag has been reverted with these as a casualty.

== Changes to 0.99.23 ==
 * Client code has undergone a refactor.  Numerous bugs have been addressed
  with these that were found while improving test coverage of this code
 * IO Block booting and automatic control has been added.  Please consult the
  partadm manpage for details
 * This release requires pybgsched-0.1.3
 * A user may voluntarily set their job's score to 0 and let it re-accrue using
  qalter --defer (Ticket #590)

== Changes to 0.99.22 ==
 * Fix for a deadlock that could occur due to a race condition in the 
  boot-reaping code.
 * Fix for a pid collision that could occur after a forced delete of a job in
  the system-script-forker component leading to a job getting stuck in the
  Job_Epilogue state.
 * python 2.7 and it's change to the XMLRPC libraries are now supported.

== Changes to 0.99.21 ==
 * Minor fix release to correct a failure mode where we can lose a process group
  if a job's boot is delayed.

== Changes to 0.99.20 ==
 * Small changeset to fix a race condition that was hanging jobs that came in
   on 0.99.19
 * Better error logging for situations where the boots and the process group
  information haven't lined up yet.  This usually happens during high-load and
  can result in block boot completion never being detected.

== Changes to 0.99.19 ==
 * Manual boot control has been added
 * major refactor of compute block rebooting

== Changes to 0.99.18 ==
 * Fix for other systems that were seeing erronious geometry data
 * SoftwareFailure now accurately reported

== Changes to 0.99.17 ==

 * Support for --geometry has been added.  This will constrain a job to run 
blocks having a specific geometry in nodes, partadm, partlist and qstat have
been updated to show this.
 * Support for passthrough blocking reservations has been added.
 * Fixed a bug where a job run outside of Cobalt would not properly block resources.
 * Removed autoreboot in the event of a block entering a control system 
  SoftwareFalure state on BG/Q.  Normal cleanup should handle this now.

== Changes to 0.99.15 ==
 * Fix for an issue that was causing partadm to malfunction when reservations
were set on the system.

== Changes to 0.99.12 ==
 * As a note, this is a release of fixes and backported features from BG/Q
 * Fixed issues that could cause a bad hardware state to be overridden as 
"blocked" allowing the block to be drained
 * Fix for a situation where a partition that is inelligible to run due to a
pending reservation is selected for draining.


src/clients/showres.py
#!/usr/bin/env python
"""
Cobalt showres command

Usage: showres [-l] [-x] [--oldts] [--version]
version: "%prog " + __revision__ + , Cobalt  + __version__

OPTIONS DEFINITIONS:

'-d','--debug',dest='debug',help='turn on communication debugging',callback=cb_debug
'-l',dest='verbose',action='store_true',help='print reservation list verbose'
'--oldts',dest='oldts',action='store_true',help='use old timestamp'
'-x',dest='really_verbose',action='store_true',help='print reservations really verbose'

"""
import logging
import math
import sys
import time
from Cobalt import client_utils
from Cobalt.client_utils import cb_debug

from Cobalt.arg_parser import ArgParse

__revision__ = '$Revision: 2154 $'
__version__ = '$Version$'

SYSMGR = client_utils.SYSMGR
SCHMGR = client_utils.SCHMGR

def mergelist(location_string, cluster):
    if not cluster:
        return location_string

    locations = location_string.split(":")

    return client_utils.merge_nodelist(locations)

def main():
    """
    showres main
    """
    # setup logging for client. The clients should call this before doing anything else.
    client_utils.setup_logging(logging.INFO)

    # list of callback with its arguments
    callbacks = [
        # <cb function>     <cb args>
        [ cb_debug        , () ] ]

    # Get the version information
    opt_def =  __doc__.replace('__revision__',__revision__)
    opt_def =  opt_def.replace('__version__',__version__)

    parser = ArgParse(opt_def,callbacks)

    parser.parse_it() # parse the command line

    if not parser.no_args():
        client_utils.logger.error("No arguments needed")
    
    if parser.options.verbose != None and parser.options.really_verbose != None:
        client_utils.logger.error('Only use -l or -x not both')
        sys.exit(1)

    cluster = False
    if 'cluster' in client_utils.component_call(SYSMGR, False, 'get_implementation', ()):
        cluster = True

    reservations = client_utils.component_call(SCHMGR, False, 'get_reservations', 
                                               ([{'name':'*', 'users':'*','start':'*', 'duration':'*', 'partitions':'*', 
                                                  'cycle': '*', 'queue': '*', 'res_id': '*', 'cycle_id': '*','project':'*', 
                                                  'block_passthrough':'*'}], ))

    output = []

    verbose        = False
    really_verbose = False
    header = [('Reservation', 'Queue', 'User', 'Start', 'Duration','Passthrough', 'Partitions', 'Time Remaining')]

    if parser.options.verbose:
        verbose = True
        header = [('Reservation', 'Queue', 'User', 'Start', 'Duration',
                   'End Time', 'Cycle Time', 'Passthrough', 'Partitions', 'Time Remaining')]
    elif parser.options.really_verbose:
        really_verbose = True
        header = [('Reservation', 'Queue', 'User', 'Start', 'Duration','End Time', 'Cycle Time','Passthrough','Partitions', 
                   'Project', 'ResID', 'CycleID', 'Time Remaining' )]

    for res in reservations:

        passthrough = "Allowed"
        if res['block_passthrough']:
            passthrough = "Blocked"

        start     = float(res['start'])
        duration  = float(res['duration'])
        now       = time.time()

        deltatime = now - start
        timeleft = "Not Started" if deltatime < 0.0 else client_utils.get_elapsed_time(deltatime, duration)
        timeleft = "00:00:00" if '-' in timeleft else timeleft

        # do some crazy stuff to make reservations which cycle display the 
        # "next" start time
        if res['cycle']:
            cycle = float(res['cycle'])
            periods = math.floor((now - start)/cycle)
            # reservations can't become active until they pass the start time 
            # -- so negative periods aren't allowed
            if periods < 0:
                pass
            # if we are still inside the reservation, show when it started
            elif (now - start) % cycle < duration:
                start += periods * cycle
            # if we are in the dead time after the reservation ended, show 
            # when the next one starts
            else:
                start += (periods+1) * cycle
        if res['cycle_id'] == None:
            res['cycle_id'] = '-'

        if res['cycle']:
            cycle = float(res['cycle'])
            if cycle < (60 * 60 * 24):
                cycle = "%02d:%02d" % (cycle/3600, (cycle/60)%60)
            else:
                cycle = "%0.1f days" % (cycle / (60 * 60 * 24))
        else:
            cycle = None
        dmin = (duration/60)%60
        dhour = duration/3600

        time_fmt = "%c"
        starttime = time.strftime(time_fmt, time.localtime(start))
        endtime   = time.strftime(time_fmt, time.localtime(start + duration)) 

        if parser.options.oldts == None:
            #time_fmt += " %z (%Z)"
            starttime = client_utils.sec_to_str(start)
            endtime = client_utils.sec_to_str(start + duration)

        if really_verbose:
            output.append((res['name'], res['queue'], res['users'], 
                           starttime,"%02d:%02d" % (dhour, dmin),
                           endtime, cycle, passthrough,
                           mergelist(res['partitions'], cluster), 
                           res['project'], res['res_id'], res['cycle_id'], timeleft))
        elif verbose:
            output.append((res['name'], res['queue'], res['users'], 
                           starttime,"%02d:%02d" % (dhour, dmin),
                           endtime, cycle, passthrough,
                           mergelist(res['partitions'], cluster), 
                           timeleft))
        else:
            output.append((res['name'], res['queue'], res['users'], 
                           starttime,"%02d:%02d" % (dhour, dmin), passthrough,
                           mergelist(res['partitions'], cluster), 
                           timeleft))

    output.sort( (lambda x,y: cmp( time.mktime(time.strptime(x[3].split('+')[0].split('-')[0].strip(), time_fmt)), 
                                   time.mktime(time.strptime(y[3].split('+')[0].split('-')[0].strip(), time_fmt))) ) )
    client_utils.print_tabular(header + output)

if __name__ == '__main__':
    try:
        main()
    except SystemExit:
        raise
    except Exception, e:
        client_utils.logger.fatal("*** FATAL EXCEPTION: %s ***", e)
        sys.exit(1)
